### ISSUE TITLE
Implement scipy.spatial

### ISSUE THREAD

**Issue by oleg-kachan** [google/jax#9235]
It would be nice to implement `scipy.spatial`, starting from `distance` for pairwise distance computations.

**Comment by shoyer** [comment]
Functions for computing distances would certainly be welcome, but I doubt it makes sense to implement more complex data structures like KDTree and Delaunay in JAX. These data structures probably can't be implemented efficiently in XLA due to non-static memory requirements, and hence would require specialized low-level CPU, GPU and/or TPU implementations that would be best saved for an external package.

**Comment by jakevdp** [comment]
Agreed, thanks @shoyer! I think JAX is particularly well-suited to implementing `scipy.spatial.distance` functions. The individual metric functions (like `euclidean`, `minkowski`, etc.) are straightforward, and then the pairwise wrappers (`cdist` and `pdist`) could basically be implemented directly with `vmap`. This would be a great place for a new contributor to start.

**Comment by dkrasner** [comment]
@jakevdp new to Jax dev here, but this looks like a good way to dive in. I'll work through the [scipy.spatial.distance](https://docs.scipy.org/doc/scipy/reference/spatial.distance.html) functions one at a time, add docs etc - might send a pre-PR to make sure we are all on the same page if that works for you.

**Comment by jakevdp** [comment]
@dkrasner - that sounds great! Feel free to tag me for review in the PRs, and work-in-progress PRs are just fine.

**Comment by deneganisme** [comment]
Hey @jakevdp it's been over a year since any activity on this. If there's still interest (/still a belief that this is a great place for a new contributor to start) I can open a PR that starts to tackle the metric functions (e.g. `euclidean`). Lmk 🙂

**Comment by ingmarschuster** [comment]
started an implementation, see PR

**Comment by chrisflesher** [comment]
It looks like this issue is being discouraged now, see [JEP 18137](https://jax--18137.org.readthedocs.build/en/18137/jep/18137-numpy-scipy-scope.html). To work around this I started an implementation for `scipy.spatial` here that includes some rotation and distance functions:

https://pypi.org/project/jax-scipy-spatial/

**Comment by cdelv** [comment]
Hi, just as an idea. Making a KD tree in JAX might not make much sense, but other algorithms for contact detection and nearest neighbors could be friendly to JAX, such as cell lists or spatial hashing. Could JAX have some spatial hashing nearest neighbors? Sweep and Prue also seem quite efficient for GPUs (https://mathweb.ucsd.edu/~sbuss/ResearchWeb/EnhancedSweepPrune/SAP_paper_online.pdf, https://odr.chalmers.se/server/api/core/bitstreams/da6bab77-057f-457d-b896-b53d2419dfc2/content).

**Comment by chrisflesher** [comment]
Maybe spatial hashing methods would work somehow? Maybe ANNOY would be easy to implement and effective.

It looks like the sweep and prue use octree which is probably not easy to implement in JAX.

**Comment by cdelv** [comment]
Hi @jakevdp, @dkrasner, @shoyer, @chrisflesher , and @ingmarschuster 

I made this code for a DEM simulation. If anyone finds a more performant way to do it, I'd appreciate it if you could share it. I compared this code against Jax MD, and it performs better. Albeit this is a very lacking implementation. 

```
import jax
import jax.numpy as jnp
from jax import lax
import jax.profiler
jax.profiler.start_server(9999)

import time
import pandas as pd

device = "gpu"
jax.config.update("jax_platform_name", device)

def create_neighbor_mask(n):
    cell_mask = []
    for ix in range(-n, n + 1):
        for iy in range(-n, n + 1):
            cell_mask.append([ix, iy])
    return jnp.asarray(cell_mask, dtype=int)

neighbor_cell_mask = create_neighbor_mask(1)

@jax.jit
def morton2D(x, y):
    x = jnp.uint32(x)
    y = jnp.uint32(y)
    x = (x | (x << 8)) & jnp.uint32(0x00FF00FF)
    x = (x | (x << 4)) & jnp.uint32(0x0F0F0F0F)
    x = (x | (x << 2)) & jnp.uint32(0x33333333)
    x = (x | (x << 1)) & jnp.uint32(0x55555555)
    y = (y | (y << 8)) & jnp.uint32(0x00FF00FF)
    y = (y | (y << 4)) & jnp.uint32(0x0F0F0F0F)
    y = (y | (y << 2)) & jnp.uint32(0x33333333)
    y = (y | (y << 1)) & jnp.uint32(0x55555555)
    return x | (y << 1)

@jax.jit
def CalcHash(cell_coord, NCells):
    Hash = morton2D(cell_coord[0], cell_coord[1]) # Z-orcer curve for efficient memory layout
    #Hash =  cell_coord[0] + NCells*cell_coord[1]
    return Hash

@jax.jit
def force_between(pos_i, pos_j, rad, k):
    rij = pos_i - pos_j
    r = jnp.linalg.norm(rij)
    s = jnp.maximum(0.0, 2.0 * rad/(r+1e-16) - 1)
    return k * s * rij

@jax.jit
def compute_force_cell(i, cell, positions, particleHash, CellSize, NCells):
    cell_hash = CalcHash(cell, NCells)
    start_idx = jnp.searchsorted(particleHash, cell_hash, side='left', method='scan_unrolled')
    valid_H = (start_idx < particleHash.shape[0])*(particleHash[start_idx] == cell_hash)    
    def loop_body(k):
        candidate_index = start_idx + k
        valid_candidate = valid_H * (candidate_index < particleHash.shape[0]) * (particleHash[candidate_index] == cell_hash)
        return jax.lax.cond(
            valid_candidate,
            lambda _: force_between(positions[i], positions[candidate_index], 1.0, 10000.0),
            lambda _: jnp.zeros_like(positions[i]),
            operand = None
        )
    return jax.vmap(loop_body)(jnp.arange(8)).sum(axis=0)

@jax.jit
def ComputeForces(positions, accel, particleHash, CellSize, NCells):
    def loop_body(i):
        neighbor_cells = jnp.floor_divide(positions[i], CellSize).astype(int) + neighbor_cell_mask
        return jax.vmap(lambda cell: compute_force_cell(i, cell, positions, particleHash, CellSize, NCells))(neighbor_cells).sum(axis=0)
    return jax.vmap(loop_body)(jnp.arange(len(positions)))

@jax.jit
def reflect(pos, rad, L):
    left_pen = jnp.maximum(rad - pos, 0.0)
    right_pen = jnp.maximum(pos - (L - rad), 0.0)
    return 100000.0 * left_pen - 100000.0 * right_pen

@jax.jit
def step_optimized(positions, velocities, accel, particleHash, dt, rad, CellSize, NCells, L):
    particleHash = jax.vmap(CalcHash, (0, None))(jnp.floor_divide(positions, CellSize).astype(int), NCells)
    sorted_indices = jnp.argsort(particleHash)
    positions = positions[sorted_indices]
    velocities = velocities[sorted_indices]
    particleHash = particleHash[sorted_indices]
    accel = ComputeForces(positions, accel, particleHash, CellSize, NCells)
    accel += reflect(positions, rad, L)
    velocities += dt * accel
    positions += dt * velocities
    return positions, velocities, accel


@jax.jit
def ComputeForcesNaive(positions, accel, rad, k):
    return jax.vmap(
        lambda pos_i: jax.lax.map(
            lambda pos_j: force_between(pos_i, pos_j, 1.0, 10000.0), positions, batch_size=32
        ).sum(axis=0)
    )(positions)

@jax.jit
def step_naive(positions, velocities, accel, dt, rad, L):
    accel = ComputeForcesNaive(positions, accel, rad, 10000.0)
    accel += reflect(positions, rad, L)
    velocities += dt * accel
    positions += dt * velocities
    return positions, velocities, accel


def benchmark():
    density = 0.1
    dt = 0.01
    rad = 1.0
    num_steps = 200 
    
    Ns = [2**i for i in range(16)]
    perf_optimized = []
    perf_naive = []
    
    for N in Ns:
        L = jnp.sqrt(N / density)
        CellSize = 2*rad
        NCells = int(jnp.floor(L / CellSize))
        key = jax.random.PRNGKey(0)
        positions = jax.random.uniform(key, (N, 2), minval=0.0, maxval=L)
        velocities = jax.random.uniform(key, (N, 2), minval=0.0, maxval=1.0)
        accel = jnp.zeros_like(positions)
        particleHash = jnp.zeros(N, dtype=int)
        
        # Warm-up jit compilation
        positions, velocities, accel = step_optimized(positions, velocities, accel, particleHash, dt, rad, CellSize, NCells, L)
        
        # Measure optimized performance
        start = time.time()
        for _ in range(num_steps):
            positions, velocities, accel = step_optimized(positions, velocities, accel, particleHash, dt, rad, CellSize, NCells, L)
        end = time.time()

        perf_opt = N * num_steps / (end - start)
        perf_optimized.append(perf_opt)
        
        positions = jax.random.uniform(key, (N, 2), minval=0.0, maxval=L)
        velocities = jax.random.uniform(key, (N, 2), minval=0.0, maxval=1.0)
        accel = jnp.zeros_like(positions)

        # Warm-up jit compilation
        positions, velocities, accel = step_naive(positions, velocities, accel, dt, rad, L)
        
        start = time.time()
        for _ in range(num_steps):
           positions, velocities, accel = step_naive(positions, velocities, accel, dt, rad, L)
        end = time.time()
        perf_nv = N * num_steps / (end - start)
        perf_naive.append(perf_nv)
        
        print(f"N = {N}: Optimized = {perf_opt:.1f}, Naive = {perf_nv:.1f}, speedup = {perf_opt/perf_nv:.2f}")


if __name__ == '__main__':
    benchmark()
```

It has a good speedup compared with the naive version. I'm not a JAX wizard, so this can be made faster. Due to JAX immutability, one has to assume that a maximum amount of spheres fit on a cell. To query nearest neighbors, one could introduce an artificial sphere.  This implementation is inspired by https://developer.download.nvidia.com/assets/cuda/files/particles.pdf. I also know that one can leverage the magic of the Z-order curve to get an octree "for free". 

The performance degrades once you increase the number of neighbor cells to search. Maybe using a lookup table and neighbor lookups for the hashing function could help. 

These are the performance numbers I got on my laptop (NVIDIA GeForce RTX 3050 Mobile, AMD Ryzen 9 6900HS):

```
N = 1: Optimized = 12282.2, Naive = 14550.2, speedup = 0.84
N = 2: Optimized = 22353.2, Naive = 29076.6, speedup = 0.77
N = 4: Optimized = 44310.9, Naive = 50229.7, speedup = 0.88
N = 8: Optimized = 88283.7, Naive = 117896.2, speedup = 0.75
N = 16: Optimized = 182945.2, Naive = 213467.6, speedup = 0.86
N = 32: Optimized = 347601.8, Naive = 447713.3, speedup = 0.78
N = 64: Optimized = 682833.4, Naive = 654289.8, speedup = 1.04
N = 128: Optimized = 1429597.1, Naive = 1244355.4, speedup = 1.15
N = 256: Optimized = 2853457.6, Naive = 1968831.8, speedup = 1.45
N = 512: Optimized = 5695185.6, Naive = 2870756.4, speedup = 1.98
N = 1024: Optimized = 11124984.9, Naive = 2148600.9, speedup = 5.18
N = 2048: Optimized = 22294146.4, Naive = 1975178.9, speedup = 11.29
N = 4096: Optimized = 35522386.1, Naive = 1213055.1, speedup = 29.28
N = 8192: Optimized = 45009121.6, Naive = 568055.3, speedup = 79.23
N = 16384: Optimized = 48757619.7, Naive = 286528.8, speedup = 170.17
```

**Comment by dodgebc** [comment]
Re-posting this from #9813 since it seems relevant

It's actually possible to do efficient k-nearest neighbors with a k-d tree using only XLA primitives thanks so some beautiful GPU-friendly tree algorithms [[1](https://arxiv.org/abs/2211.00120), [2](https://arxiv.org/abs/2210.12859)] which are implemented in [cudaKDTree](https://github.com/ingowald/cudaKDTree).

Here's a lightweight JAX version that I hope will be useful to some folks!
https://github.com/dodgebc/jaxkd

**Comment by mattjj** [comment]
@dodgebc that’s awesome, thanks for sharing!

**Comment by shoyer** [comment]
@dodgebc  Really awesome indeed, well done!
